### PR TITLE
Ergo LM instantiation

### PR DESCRIPTION
**Pull request by oskin1**

### CODE REVIEW

File: spectrum-offchain-lm/src/data/pool.rs [hunk 1]
```diff
@@ -0,0 +1,583 @@
...
+        let bundle = StakingBundleProto {
+            bundle_key_id: bundle_key.to_asset(),
+            pool_id: next_pool.pool_id,
+            vlq: release_vlq,
+            tmp: release_tmp,
+            redeemer_prop: deposit.redeemer_prop.clone(),
+            erg_value: NanoErg::from(BoxValue::SAFE_USER_MIN),
+        };
+        let user_output = DepositOutput {
+            bundle_key,
```
Comment (kettlebell) [inline spectrum-offchain-lm/src/data/pool.rs:115]: So `user_output` will have MAX_VALUE of BundleKey tokens, but doesn't this conflict with this line of the contract: https://github.com/spectrum-finance/ergo-dex/blob/ad099cc64fb35b1c7a9a72d97631a8776281ed75/contracts/lqmining/simple/Deposit.sc#L48?

**Review by kettlebell — COMMENTED** [review]
Review up to Deposit action

File: spectrum-offchain-lm/src/data/pool.rs [hunk 2]
```diff
@@ -0,0 +1,552 @@
...
+        .unwrap();
+        let registers = NonMandatoryRegisters::try_from(
+            vec![
+                Constant::from(<Vec<i32>>::from(self.conf)),
+                Constant::from(self.conf.program_budget.amount as i64),
+                Constant::from(self.max_error as i64),
+            ]
+            .into_iter()
+            .chain(
+                self.epoch_ix
```
[the remark this answers is not in the document]
Comment (oskin1) [reply spectrum-offchain-lm/src/data/pool.rs:333]: Right, we support only self hosted version initially

File: spectrum-offchain-lm/src/data/order.rs [hunk 3]
```diff
@@ -0,0 +1,577 @@
...
+        RunOrderError<Self>,
+    > {
+        let unwrapped_bundles = bundles.clone().into_iter().map(|AsBox(_, b)| b).collect();
+        match pool.distribute_rewards(unwrapped_bundles, funding.clone().map(|AsBox(_, f)| f)) {
+            Ok((next_pool, next_bundles, next_funding, rewards)) => {
+                let outputs = TxIoVec::from_vec(
+                    vec![next_pool.clone().into_candidate(ctx.height)]
+                        .into_iter()
+                        .chain(
+                            next_funding
```
[the remark this answers is not in the document]
Comment (oskin1) [reply spectrum-offchain-lm/src/data/order.rs:103]: In context of "rewarding" operation impl indexing beyond pool (INPUTS(0), OUTPUTS(0)) doesn't matter. Other indexes are dynamic and are written into ContextExtensions when the order is evaluated.

File: spectrum-offchain-lm/src/data/bundle.rs [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+}
+
+impl IntoBoxCandidate for StakingBundleProto {
+    fn into_candidate(self, height: u32) -> ErgoBoxCandidate {
+        let tokens = BoxTokens::from_vec(vec![
+            Token::from(self.vlq),
+            Token::from(self.tmp),
+            Token {
+                token_id: self.bundle_key_id.token_id,
+                amount: TokenAmount::try_from(BUNDLE_KEY_AMOUNT).unwrap(),
```
[the remark this answers is not in the document]
Comment (oskin1) [reply spectrum-offchain-lm/src/data/bundle.rs:54]: Fixed

File: spectrum-offchain/src/executor.rs [hunk 5]
```diff
@@ -1,112 +1,137 @@
...
+pub fn executor_stream<'a, TExecutor: Executor + 'a>(executor: TExecutor) -> impl Stream<Item = ()> + 'a {
     let executor = Arc::new(Mutex::new(executor));
-    stream::iter(0..).then(move |_| {
+    stream::unfold((), move |_| {
         let executor = executor.clone();
         async move {
-            let mut executor_quard = executor.lock();
-            executor_quard.execute_next().await
+            trace!(target: "offchain_lm", "Trying to execute next order ..");
+            let mut executor_quard = executor.lock().await;
```
Comment (kettlebell) [inline spectrum-offchain/src/executor.rs:129]:
```suggestion
            let mut executor_guard = executor.lock().await;
```

File: spectrum-offchain-lm/src/data/order.rs [hunk 6]
```diff
@@ -0,0 +1,608 @@
...
+                .ok()?;
+                let expected_num_epochs = bx
+                    .ergo_tree
+                    .get_constant(13)
+                    .ok()??
+                    .v
+                    .try_extract_into::<i32>()
+                    .ok()?;
+                let lq = TypedAssetAmount::<Lq>::from_token(tokens.get(0)?.clone());
+                return Some(Deposit {
```
[the remark this answers is not in the document]
Comment (oskin1) [reply spectrum-offchain-lm/src/data/order.rs:330]: RefundPk is not required for execution.

File: spectrum-offchain-lm/src/data/order.rs [hunk 7]
```diff
@@ -0,0 +1,608 @@
...
+        }
+        None
+    }
+}
+
+#[derive(Debug, Eq, PartialEq, Clone, Serialize, Deserialize)]
+#[serde(from = "RawRedeem")]
+#[serde(into = "RawRedeem")]
+pub struct Redeem {
+    pub order_id: OrderId,
```
[the remark this answers is not in the document]
Comment (oskin1) [reply spectrum-offchain-lm/src/data/order.rs:348]: Same here

**Pull request merged**